core/metadata.cc

```diff
@@ -260,123 +290,75 @@ static void FillOffsetArrays() {
 
       if (!invalid && offset >= 0) {
         for (int l = 0; l < size; l++) {
-          m->scope_components[offset + l] = i;
+          module_components_[m][offset + l] = i;
         }
       }
     }
   }
 }
 
-static int Disjoint(int scope1, int scope2) {
-  const struct scope_component &comp1 = scope_components[scope1];
-  const struct scope_component &comp2 = scope_components[scope2];
-
-  for (const auto &i : comp1.modules) {
-    for (const auto &j : comp2.modules) {
-      if (i == j) {
-        return 0;
-      }
-    }
-  }
-  return 1;
-}
-
-static mt_offset_t NextOffset(mt_offset_t curr_offset, int8_t size) {
-  uint32_t overflow;
-  int8_t rounded_size;
-
-  rounded_size = align_ceil_pow2(size);
-
-  if (curr_offset % rounded_size != 0) {
-    curr_offset = ((curr_offset / rounded_size) + 1) * rounded_size;
-  }
-
-  overflow = (uint32_t)curr_offset + (uint32_t)size;
-
-  return overflow > kMetadataTotalSize ? kMetadataOffsetNoSpace : curr_offset;
-}
-
-static void AssignOffsets() {
+void Pipeline::AssignOffsets() {
   mt_offset_t offset = 0;
-  std::priority_queue<struct scope_component *,
-                      std::vector<struct scope_component *>,
-                      std::function<bool(const struct scope_component *,
-                                         const struct scope_component *)>>
+  std::priority_queue<
+      ScopeComponent *, std::vector<ScopeComponent *>,
+      std::function<bool(const ScopeComponent *, const ScopeComponent *)>>
```

Comment:
Just to clarify -- it wouldn't be an overload in the class but rather a binary function implemented as a class with its operator defined.